src/enhancers/convert_frequencies_to_stoptimes.rs

```diff
@@ -0,0 +1,160 @@
+    objects::{Calendar, Frequency, StopTime, Time, VehicleJourney},
+    Result,
+};
+use anyhow::anyhow;
+use skip_error::skip_error_and_warn;
+use std::collections::{BTreeSet, HashMap};
+use tracing::warn;
+use typed_index_collection::CollectionWithId;
+
+/// Convert all frequencies to stoptimes
+/// by creating new duplicated vehicle_journeys, calendars and comments if necessary
+/// and remove all frequencies from Collections
+pub fn convert_all_frequencies_to_stoptimes(collections: &mut Collections) -> Result<()> {
+    let frequencies = collections.frequencies.take();
+    convert_frequencies_to_stoptimes(collections, frequencies)
+}
+
+pub fn convert_frequencies_to_stoptimes(
+    collections: &mut Collections,
+    frequencies: Vec<Frequency>,
+) -> Result<()> {
+    let mut trip_id_sequence: HashMap<String, u32> = HashMap::new();
+    let mut new_vehicle_journeys: Vec<VehicleJourney> = vec![];
+
+    for frequency in &frequencies {
+        if frequency.start_time == frequency.end_time {
+            warn!(
+                "frequency for trip {:?} has same start and end time",
+                frequency.vehicle_journey_id
+            );
+            continue;
+        }
+        let corresponding_vj = skip_error_and_warn!(collections
+            .vehicle_journeys
+            .get(&frequency.vehicle_journey_id)
+            .cloned()
+            .ok_or_else(|| anyhow!(
+                "frequency mapped to an unexisting trip {:?}",
+                frequency.vehicle_journey_id
+            )));
+        let mut start_time = frequency.start_time;
+        let mut arrival_time_delta = match corresponding_vj.stop_times.iter().min() {
+            None => {
+                warn!(
+                    "frequency mapped to trip {:?} with no stop_times",
+                    frequency.vehicle_journey_id
+                );
+                continue;
+            }
+            Some(st) => st.arrival_time,
+        };
+        while start_time < frequency.end_time {
+            trip_id_sequence
+                .entry(frequency.vehicle_journey_id.clone())
+                .and_modify(|counter| *counter += 1)
+                .or_insert(0);
+            let generated_trip_id = format!(
+                "{}-{}",
+                frequency.vehicle_journey_id, trip_id_sequence[&frequency.vehicle_journey_id]
+            );
+            // the following handles generated trip starting after midnight, we need to generate a
+            // new service in case the next day is not covered
+            let service_id = if start_time.hours() >= 24 {
+                let nb_days = start_time.hours() / 24;
+                let service = collections
+                    .calendars
+                    .get(&corresponding_vj.service_id)
+                    .cloned()
+                    .unwrap();
+                let new_service_id = format!("{}:+{}days", service.id, nb_days);
+                if !collections.calendars.contains_id(&new_service_id) {
```

Comment:
In the `else` (the service ID already exists), should we check that the service is actually the one we want?

Reply:
I don't see breach in this (historical old) part :thinking:

Reply:
I believe I messed up my reading with the multiple nesting. Seems fine.